### PR TITLE
New command: qmk console

### PR DESCRIPTION
**Pull request by skullydazed**
## Description

This adds a `qmk console` command. The initial work was done by @zvecr to connect to a single keyboard, I expanded it out to connect to all (or one) keyboard. It also shows the connected bootloaders, much in the same way QMK Toolbox does.

## Types of Changes

- [x] Core
- [x] New feature
- [x] Documentation

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: lib/python/qmk/cli/console.py [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+        ('1EAF', '0003'): 'STM32duino Bootloader',
+        ('16C0', '05DC'): 'USBaspLoader',
+        ('1C11', 'B007'): 'Kiibohd DFU',
+        ('03EB', '2FEF'): 'ATmega16U2 DFU',
+        ('03EB', '2FF0'): 'ATmega32U2 DFU',
+        ('03EB', '2FF3'): 'ATmega16U4 DFU',
+        ('03EB', '2FF4'): 'ATmega32U4 DFU',
+        ('03EB', '2FF9'): 'AT90USB64 DFU',
+        ('03EB', '2FFA'): 'AT90USB162 DFU',
+        ('03EB', '2FFB'): 'AT90USB128 DFU'
```
Comment (fauxpark) [inline lib/python/qmk/cli/console.py:40]: Missing Caterina and SAM-BA here.

File: util/install/arch.sh [hunk 2]
```diff
@@ -4,13 +4,13 @@ _qmk_install() {
...
 
     sudo pacman --needed  --noconfirm -S \
-        base-devel clang diffutils gcc git unzip wget zip \
-        python-pip \
-        avr-binutils \
-        arm-none-eabi-binutils arm-none-eabi-gcc arm-none-eabi-newlib \
-        avrdude dfu-programmer dfu-util
+        base-devel clang diffutils gcc git unzip wget zip python-pip \
+	avr-binutils arm-none-eabi-binutils arm-none-eabi-gcc \
+	arm-none-eabi-newlib avrdude dfu-programmer dfu-util
```
Comment (fauxpark) [inline util/install/arch.sh:9]:
```suggestion
        avr-binutils arm-none-eabi-binutils arm-none-eabi-gcc \
        arm-none-eabi-newlib avrdude dfu-programmer dfu-util
```

File: util/install/debian.sh [hunk 3]
```diff
@@ -13,10 +13,9 @@ _qmk_install() {
 
     sudo apt-get -yq install \
         build-essential clang-format diffutils gcc git unzip wget zip \
-        python3-pip \
-        binutils-avr gcc-avr avr-libc \
-        binutils-arm-none-eabi gcc-arm-none-eabi libnewlib-arm-none-eabi \
-        avrdude dfu-programmer dfu-util teensy-loader-cli libusb-dev
+	python3-pip binutils-avr gcc-avr avr-libc binutils-arm-none-eabi \
+	gcc-arm-none-eabi libnewlib-arm-none-eabi avrdude dfu-programmer \
+	dfu-util teensy-loader-cli libhidapi-hidraw0
```
Comment (fauxpark) [inline util/install/debian.sh:18]:
```suggestion
        python3-pip binutils-avr gcc-avr avr-libc binutils-arm-none-eabi \
        gcc-arm-none-eabi libnewlib-arm-none-eabi avrdude dfu-programmer \
        dfu-util teensy-loader-cli libhidapi-hidraw0
```

File: util/install/fedora.sh [hunk 4]
```diff
@@ -5,11 +5,10 @@ _qmk_install() {
...
     sudo dnf -y install \
-        clang diffutils git gcc glibc-headers kernel-devel kernel-headers make unzip wget zip \
-        python3 \
-        avr-binutils avr-gcc avr-libc \
-        arm-none-eabi-binutils-cs arm-none-eabi-gcc-cs arm-none-eabi-newlib \
-        avrdude dfu-programmer dfu-util libusb-devel
+        clang diffutils git gcc glibc-headers kernel-devel kernel-headers \
+	make unzip wget zip python3 avr-binutils avr-gcc avr-libc \
+	arm-none-eabi-binutils-cs arm-none-eabi-gcc-cs arm-none-eabi-newlib \
+	avrdude dfu-programmer dfu-util hidapi
```
Comment (fauxpark) [inline util/install/fedora.sh:11]:
```suggestion
        make unzip wget zip python3 avr-binutils avr-gcc avr-libc \
        arm-none-eabi-binutils-cs arm-none-eabi-gcc-cs arm-none-eabi-newlib \
        avrdude dfu-programmer dfu-util hidapi
```

File: util/install/msys2.sh [hunk 5]
```diff
@@ -9,11 +9,10 @@ _qmk_install() {
...
     pacboy sync --needed --noconfirm --disable-download-timeout \
-        base-devel: toolchain:x clang:x git: unzip: \
-        python3-pip:x \
-        avr-binutils:x avr-gcc:x avr-libc:x \
-        arm-none-eabi-binutils:x arm-none-eabi-gcc:x arm-none-eabi-newlib:x \
-        avrdude:x bootloadhid:x dfu-programmer:x dfu-util:x teensy-loader-cli:x
+        base-devel: toolchain:x clang:x git: unzip: python3-pip:x \
+	avr-binutils:x avr-gcc:x avr-libc:x arm-none-eabi-binutils:x \
+	arm-none-eabi-gcc:x arm-none-eabi-newlib:x avrdude:x bootloadhid:x \
+	dfu-programmer:x dfu-util:x teensy-loader-cli:x hidapi:x
```
Comment (fauxpark) [inline util/install/msys2.sh:15]:
```suggestion
        avr-binutils:x avr-gcc:x avr-libc:x arm-none-eabi-binutils:x \
        arm-none-eabi-gcc:x arm-none-eabi-newlib:x avrdude:x bootloadhid:x \
        dfu-programmer:x dfu-util:x teensy-loader-cli:x hidapi:x
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: lib/python/qmk/cli/console.py (hunk 1 above)
Comment (fauxpark) [reply lib/python/qmk/cli/console.py:40]: https://docs.qmk.fm/#/driver_installation_zadig?id=list-of-known-bootloaders

Comment (fauxpark) [reply lib/python/qmk/cli/console.py:40]: SAM-BA is `03EB:6124`

Comment (skullydazed) [reply lib/python/qmk/cli/console.py]: I've added these in ab0d7a378.

**Review by fauxpark — APPROVED** [review]

**Review by drashna — APPROVED** [review]
Tested on macOS, works well.

File: util/udev/50-qmk.rules [hunk 6]
```diff
@@ -60,3 +60,6 @@ SUBSYSTEMS=="usb", ATTRS{idVendor}=="239a", ATTRS{idProduct}=="000e", TAG+="uacc
 SUBSYSTEMS=="usb", ATTRS{idVendor}=="2a03", ATTRS{idProduct}=="0036", TAG+="uaccess", ENV{ID_MM_DEVICE_IGNORE}="1"
 ### Micro
 SUBSYSTEMS=="usb", ATTRS{idVendor}=="2a03", ATTRS{idProduct}=="0037", TAG+="uaccess", ENV{ID_MM_DEVICE_IGNORE}="1"
+
+# hid_listen
+KERNEL=="hidraw*", MODE="0660", GROUP="plugdev", TAG+="uaccess", TAG+="udev-acl"
```
Comment (t-8ch) [inline util/udev/50-qmk.rules:65]: Is there a concrete reason for `GROUP=plugdev`?
It is not set for the other rules and is less secure than `TAG=uaccess`

File: lib/python/qmk/cli/console.py [hunk 7]
```diff
@@ -0,0 +1,302 @@
+"""Acquire debugging information from usb hid devices
+
+cli implementation of https://www.pjrc.com/teensy/hid_listen.html
+"""
+from pathlib import Path
+from threading import Thread
+from time import sleep, strftime
+
+import hid
+import usb.core
```
Comment (t-8ch) [inline lib/python/qmk/cli/console.py:10]: This introduces a hard dependency on the `hid` and `pyusb` packages .
Without them, the `qmk` command can not be used at all anymore.
Was this intentional?

Comment (t-8ch) [reply lib/python/qmk/cli/console.py:10]: Fix submitted in https://github.com/qmk/qmk_firmware/pull/12978

File: util/udev/50-qmk.rules (hunk 6 above)
Comment (skullydazed) [reply util/udev/50-qmk.rules:65]: I don't remember where I found that, but it was in the original rule I copied. I didn't have a specific reason to add that, if there's a better configuration that could be changed.

File: lib/python/qmk/cli/console.py (hunk 7 above)
Comment (skullydazed) [reply lib/python/qmk/cli/console.py:10]: Yes, that was intentional, those dependencies are now in `requirements.txt` and the necessary OS packages are in the install scripts.

File: util/udev/50-qmk.rules (hunk 6 above)
Comment (t-8ch) [reply util/udev/50-qmk.rules:65]: `TAG+="uaccess"` should be enough. It makes sure that the currently locally present user has access to the device node.
`udev-acl` may be the old version of that.

Given that all the other rules in this file only use `TAG+="uaccess"` that probably also works here.

Comment (t-8ch) [reply util/udev/50-qmk.rules:65]: When `GROUP=` is removed, `MODE=` is also not necessary anymore.

Comment (skullydazed) [reply util/udev/50-qmk.rules:65]: Would you like to submit that PR? You seem to know a lot more about udev rules than I do and it's always good to get more knowledgeable hands on it.

Comment (t-8ch) [reply util/udev/50-qmk.rules:65]: Sure

**Pull request merged**